vllm/v1/cudagraph_dispatcher.py

```diff
@@ -113,21 +162,19 @@ def dispatch(
         if not self.keys_initialized:
             return CUDAGraphMode.NONE, None
 
-        non_uniform_key = batch_descriptor.non_uniform
-        # if a batch use cascade attention, bypass checking full cudagraphs
-        if not use_cascade_attn:
-            # check if key exists for full cudagraph
-            if batch_descriptor in self.cudagraph_keys[CUDAGraphMode.FULL]:
-                return CUDAGraphMode.FULL, batch_descriptor
+        num_tokens, uniform_decode = (
+            batch_descriptor.num_tokens,
+            batch_descriptor.uniform_decode,
+        )
 
-            # otherwise, check if non-uniform key exists
-            if non_uniform_key in self.cudagraph_keys[CUDAGraphMode.FULL]:
-                return CUDAGraphMode.FULL, non_uniform_key
+        candidates = [(CUDAGraphMode.FULL, (num_tokens, uniform_decode))]
+        if uniform_decode:
+            candidates.append((CUDAGraphMode.FULL, (num_tokens, False)))
+        candidates.append((CUDAGraphMode.PIECEWISE, (num_tokens, False)))
 
-        # also check if non-uniform key exists for more "general"
-        # piecewise cudagraph
-        if non_uniform_key in self.cudagraph_keys[CUDAGraphMode.PIECEWISE]:
-            return CUDAGraphMode.PIECEWISE, non_uniform_key
+        for mode, key in candidates:
```

Comment:
I think this makes the logic a bit more confusing so we should add comments and add to the dispatching docs. What we're saying here is that if we have a uniform cudagraph that cannot handle the required number of requests, we downgrade to piecewise? Will that happen or do we make sure it doesn't ahead of time? Just worried about this kind of "hidden fallback"